File: day_1/day1.py
from pathlib import Path
from itertools import combinations
from math import prod
import timeit
from collections import defaultdict

def get_input():
    input_path = Path(__file__).parent / "input.txt"
    data = set(int(line) for line in input_path.open())
    return data

def part1(data: set) -> int:
    for tuple in combinations(data, 2):
        if sum(tuple) == 2020:
            return prod(tuple)

def part2(data: set) -> int:
    values = defaultdict()
    for tuple in combinations(data, 2):
        values[sum(tuple)] = prod(tuple)
    for number in data:
        try:
            return values[2020-number]*number
        except KeyError:
            pass


if __name__ == "__main__":
    data = get_input()
    print(part1(data))
    print(timeit.timeit(stmt="part1(data)", number=5, setup="from __main__ import part1, get_input; data=get_input();")/5)
    print(part2(data))
    print(timeit.timeit(stmt="part2(data)", number=5, setup="from __main__ import part2, get_input; data=get_input();")/5)

File: day_1/day1.nim
import strutils
import tables
from math import prod
from algorithm import product
import os

proc get_input(): seq[int] =
    var input_file = currentSourcePath().parentDir() & r"\input.txt"
    var f = readFile(input_file).splitLines()
    var data: seq[int] = @[]
    for line in f:
        data.add(line.parseInt)
    return data

proc part1(data: seq[int]): int =
    var dict = initTable[int, int]()

    for num in data:
        dict[2020-num] = num

    for num in data:
        if num in dict:
            return num * dict[num]

proc part2(data: seq[int]): int =
    var dict = initTable[int, seq[int]]()

    for tup in product(@[data,data]):
        var key = 2020-tup[0]-tup[1]
        dict[key] = tup

    for num in data:
        if num in dict:
            return num * prod(dict[num])

var data = get_input()
echo "Part1: ", part1(data)
echo "Part2: ", part2(data)







